Refactor ImageDataGenerator

vkk800 opened the pull request:
This PR refactorizes ImageDataGenerator so that the random parameters of a transformation are generated separately from applying the transformation. The generation of the parameters happens in `get_random_transform()` function which outputs a dictionary of transformation parameters randomly chosen. The transform can then be applied in a new `apply_transform()` function.

My aim is to eventually enable more generalized usage for ImageDataGenerator; for example, transformation of segmentation coordinates in an image segmentation problem along with the image itself. Since this is going to be quite a big upgrade, I'm doing it in smaller steps at a time to be sure not to break anything too horribly. This refactorization is the first step.

Some design choices I have made here:

* Representing a transformation as a dictionary of transformation parameters and values corresponding to each of them. Other choice would be just a list or tuple of the values, but I think this way is more generalizable in the future if we want e.g. custom transformation functions (for image segmentation coordinates)

* Keeping all the old functions (`random_transform()` for instance) even though they are not strictly needed. This is not to break any existing functionality anywhere else. Now the old functions just call the new functions.

To elaborate on my future vision on the ImageDataGenerator: my plan is that eventually it would be able to process a list of inputs and a list of outputs, some of which might be images and some which might be something else. It would also take, as a constructor parameter, a list of transformations that need to applied to each of the inputs and outputs. The manner in which a given transformation acts on a given type of input is specified by giving separate set of transformation functions for each input/output. Some of them can be in-build (e.g. what we have now) or user defined.

Please comment on my plans now so that I don't waste my time if these changes are not desired or someone else is already working on a completely different plan.

brge17 commented:
Drive by comment:

+ 1

Strong data typing and known tasks (classification, detection, segmentation) allow you to make a more general purpose image generator. As you point out transformation such as random translation would move the bounding boxes in detection and the output image in segmentation. However, this is an implementation detail as the user is telling you what they want to do. Adding additional tasks becomes as simple as writing extensions to the augmentation function and new data types.

Designs like these allow you to do any arbitrary COCO task with the same generator.

fchollet commented:
@Dref360 would you have the bandwidth to take a look?

Dref360 commented:
Sure, it will be done by the end of the week.
Sorry kind of a crazy month.

vkk800 commented:
@Dref360 Such functions could be created for situations when the user wants to apply just one transformation, but `apply_matrix_transform()` should still be the primary way of doing the transformations. The reason is that it is more efficient to first combine all the affine transformations into one 3x3 transformation matrix and then apply it to the image rather than apply all the transformations separately. Also, I think that all the preprocessing functions are used mainly through `ImageDataGenerator` class so the use for easy-to-use image transformation functions is probably pretty marginal.

Alternative way would be to have `apply_affine_transformation()` function function that takes all the affine transformation parameters as argument and then constructs the matrix and applies it. Now that I think of it, this might be slightly nicer way of doing things. I will make a new commit to change it, but this will take a few days because my schedule for the next days is a bit busy.

vkk800 commented:
@Dref360 I just added a new commit that:

* Adds `apply_affine_transform()` function.

* Removes `apply_matrix_transform()` function

* Updates all the other functions that do affine transformations to use `apply_affine_transform()`.

I also fixed the small issue pointed out by @fchollet.

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -972,55 +1018,99 @@ def random_transform(self, x, seed=None):
+
+        return transform_parameters
+
+    def apply_transform(self, x, transform_parameters):
```
This method is not usable by users in the general case since `transform_parameters` is not really documented. Please make this method private.

vkk800 replied:
However, the functions is used outside `ImageDataGenerator` class by `NumpyArrayIterator`. Also, it is conceivable (although not perhaps the most common use case) that the user wants to use this function directly for something. This is why I would rather document the transformations than make the function private. The same applies for `get_random_transform`.

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -907,30 +956,27 @@ def standardize(self, x):
 
-    def random_transform(self, x, seed=None):
-        """Randomly augments a single image tensor.
+    def get_random_transform(self, img_shape, seed=None):
```
Please make this method private.

vkk800 commented:
I made a new commit that documents the transformation parameter dictionary in `apply_transform()`. As I commented above, I think this is better than making it private since it is used outside the class by `NumpyArrayIterator`.

(Another cosmetic fix I made was removing a few unnecessary lines of code that I forgot to remove earlier.)
